tests/cpp/test_multidevice_transformer.cpp

```diff
@@ -293,57 +446,167 @@ std::vector<TensorView*> mlp_backwards(
   constexpr double kScale = 1.0 / (1.0 - kDropoutProb);
-  Val* dscale = IrBuilder::create<Val>(kScale);
-  TensorView* dropout_grad = dropout_backward(grad, mask, dscale);
-  TensorView* dropout_grad_q = castOp(dtype, dropout_grad);
-
-  TensorView* w1_t = transpose(w1, 1, 2);
-  TensorView* matmul1_grad_x = matmul(dropout_grad_q, w1_t);
-  TensorView* grad_t = transpose(dropout_grad_q, 0, 1);
-  TensorView* matmul1_grad_w_t = matmul(grad_t, gelu);
-  TensorView* matmul1_grad_w = transpose(matmul1_grad_w_t, 1, 2);
-  TensorView* matmul1_grad_b = sum(dropout_grad, {0});
-
-  TensorView* matmul1_grad_x_ = castOp(DataType::Float, matmul1_grad_x);
+  Val* dropout_scale = IrBuilder::create<Val>(kScale);
+  TensorView* dropout_grad = dropout_backward(grad, mask, dropout_scale);
+
+  auto linear1_grads = sharded_linear_backwards(gelu, w1, dropout_grad, dtype);
+
+  TensorView* matmul1_grad_x_ = castOp(DataType::Float, linear1_grads.grad_x);
   TensorView* gelu_grad = tanh_gelu_backward(matmul1_grad_x_, linear0);
-  TensorView* gelu_grad_f = castOp(dtype, gelu_grad);
 
-  TensorView* w0_t = transpose(w0, 1, 2);
-  TensorView* matmul0_grad_x_partial = matmul(gelu_grad_f, w0_t);
-  TensorView* matmul0_grad_x = sum(matmul0_grad_x_partial, {0}); // allreduce
-  TensorView* grad_gelu_t = transpose(gelu_grad_f, 1, 2);
-  TensorView* matmul0_grad_w_t = matmul(grad_gelu_t, x);
-  TensorView* matmul0_grad_w = transpose(matmul0_grad_w_t, 1, 2);
-  TensorView* matmul0_grad_b = sum(gelu_grad, {1});
+  auto linear0_grads = linear_backwards(x, w0, gelu_grad, dtype);
 
   // Manaul sharding annotations
   for (auto tv :
-       {x, grad, mask, dropout_grad, matmul1_grad_b, matmul0_grad_x}) {
+       {x,
+        grad,
+        mask,
+        dropout_grad,
+        linear1_grads.grad_b,
+        linear0_grads.grad_x}) {
     tv->setDeviceMesh(mesh);
   }
 
   for (auto tv :
        {w0,
         b0,
         w1,
-        matmul1_grad_w,
+        linear1_grads.grad_x,
+        linear1_grads.grad_w,
         gelu_grad,
-        matmul0_grad_w,
-        matmul0_grad_b}) {
+        linear0_grads.grad_w,
+        linear0_grads.grad_b}) {
     tv->setDeviceMesh(mesh);
     tv->axis(0)->parallelize(ParallelType::DIDx);
   }
 
   std::vector<TensorView*> outputs = {
       dropout_grad,
-      matmul1_grad_w,
-      matmul1_grad_b,
+      linear1_grads.grad_w,
+      linear1_grads.grad_b,
       gelu_grad,
-      matmul0_grad_w,
-      matmul0_grad_b,
-      matmul0_grad_x};
+      linear0_grads.grad_w,
+      linear0_grads.grad_b,
+      linear0_grads.grad_x};
   return outputs;
 }
+
+std::vector<TensorView*> mha_backwards(
+    TensorView* x,
+    TensorView* w0,
+    TensorView* b0,
+    TensorView* w1,
+    TensorView* mask,
+    TensorView* sdpa_output,
+    TensorView* sdpa_log_sumexp,
+    TensorView* sdpa_seed,
+    TensorView* sdpa_offset,
+    TensorView* grad,
+    const DeviceMesh& mesh,
+    DataType dtype) {
+  // Recompute: Linear 1, QKV vectors
```

Comment:
```suggestion
  // Recompute: Linear 0, QKV vectors
```
It looks like you are recomputing linear 0 instead. Correct?